### PR TITLE
Add a `signals-based-traps` Cargo compile-time feature

### PR DESCRIPTION
**Pull request by alexcrichton**
This PR adds a new `signals-based-traps` Cargo compile-time feature which mirrors the `Config::signals_based_traps` runtime configuration. Disabling this Cargo feature enables reducing Wasmtime's platform requirements to remove the need for virtual memory and trap handlers for example. That means that the underlying platform requirements are just setjmp/longjmp and a single pointer of thread-local storage. This should make it significantly easier to port Wasmtime to platforms that may not have virtual memory in the future.

I'll note that one of the main purposes of this PR isn't quite realized just yet which is to run Pulley on embedded systems. Pulley isn't finished, however, so this is more setting the stage for that to be possible in the future than enabling it right this red-hot-second.

### CODE REVIEW

File: crates/wasmtime/Cargo.toml [hunk 1]
```diff
@@ -315,3 +334,15 @@ memory-protection-keys = ["pooling-allocator"]
...
+
+# Gates compile-time support for host signals-based-traps.
+#
+# Traps based on signals, such as SIGSEGV, are useful for accelerating
+# WebAssembly by removing explicit checks and letting the hardware deliver
+# signals instead. This feature is enabled by default and gates a number
+# of implementations within Wasmtime that may rely on virtual memory, for
+# example. Embedded systems or smaller systems may wish to disable this feature
+# to reduce the runtime requirements of Wasmtime.
+signals-based-traps = [
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply crates/wasmtime/Cargo.toml:346]: Agreed yeah, I added that in bb8fa843139ac84247e0883908a9da429ef5b7d5

File: crates/wasmtime/src/runtime/vm/mmap_vec.rs [hunk 2]
```diff
@@ -15,32 +16,53 @@ use std::{fs::File, path::Path};
 ///
 /// An `MmapVec` is an owned value which means that owners have the ability to
 /// get exclusive access to the underlying bytes, enabling mutation.
+///
+/// TODO: rename this type and reword docs now that this isn't always an mmap.
```
Comment (pchickey) [inline crates/wasmtime/src/runtime/vm/mmap_vec.rs:20]: TODO

**Review by pchickey — APPROVED** [review]
Pretty cool!

**Comment by alexcrichton** [comment]
I've pushed some extra refactoring of `MmapVec` as well to handle the internals a bit better given today's requirements.

**Pull request merged**